cartographer/mapping_2d/pose_graph/optimization_problem.cc

```diff
@@ -52,6 +57,63 @@ transform::Rigid2d ToPose(const std::array<double, 3>& values) {
   return transform::Rigid2d({values[0], values[1]}, values[2]);
 }
 
+// Selects a trajectory node closest in time to the landmark observation and
+// applies a relative transform from it.
+transform::Rigid3d GetInitialLandmarkPose(
+    const LandmarkNode::LandmarkObservation& observation,
+    const NodeData& prev_node, const NodeData& next_node) {
+  const NodeData& closest_node =
+      observation.time - prev_node.time < next_node.time - observation.time
+          ? prev_node
+          : next_node;
+  return transform::Embed3D(closest_node.pose) *
+         transform::Rigid3d::Rotation(closest_node.gravity_alignment) *
+         observation.landmark_to_tracking_transform;
+}
+
+void AddLandmarkCostFunctions(
+    const std::map<std::string, LandmarkNode>& landmark_nodes,
+    const mapping::MapById<mapping::NodeId, NodeData>& node_data,
+    mapping::MapById<mapping::NodeId, std::array<double, 3>>* C_nodes,
+    std::map<std::string, CeresPose>* C_landmarks, ceres::Problem* problem) {
+  for (const auto& landmark_node : landmark_nodes) {
+    for (const auto& observation : landmark_node.second.landmark_observations) {
+      const std::string& landmark_id = landmark_node.first;
+      // Find the trajectory nodes before and after the landmark observation.
+      auto next =
+          node_data.lower_bound(observation.trajectory_id, observation.time);
+      // The landmark observation was made before the trajectory was created.
```

Comment:
If the landmark was observed somewhere between the first and the second node, it will be dropped here.

How about
```c++
if (observation.time < 
    node_data.BeginOfTrajectory(observation.trajectory_id)->data.time) {
  continue;
}
````
?

Reply:
Good catch, it is very unlikely though, but should be fixed. Thanks for fixing it in #901, btw.